singularity-eos/eos/modifiers/minimum_energy.hpp

```diff
@@ -0,0 +1,378 @@
+//------------------------------------------------------------------------------
+
+#ifndef _SINGULARITY_EOS_EOS_MINIMUM_ENERGY_
+#define _SINGULARITY_EOS_EOS_MINIMUM_ENERGY_
+
+#include "stdio.h"
+#include <cstdlib>
+#include <iostream>
+
+#include <ports-of-call/portability.hpp>
+#include <singularity-eos/base/eos_error.hpp>
+#include <singularity-eos/eos/eos_base.hpp>
+
+namespace singularity {
+
+using namespace eos_base;
+
+template <typename T>
+class MinimumEnergy : public EosBase<MinimumEnergy<T>> {
+ public:
+  // Generic functions provided by the base class. These contain
+  // e.g. the vector overloads that use the scalar versions declared
+  // here We explicitly list, rather than using the macro because we
+  // overload some methods.
+
+  // TODO(JMM): The modifier EOS's should probably call the specific
+  // sub-functions of the class they modify so that they can leverage,
+  // e.g., an especially performant or special version of these
+  using EosBase<MinimumEnergy<T>>::TemperatureFromDensityInternalEnergy;
+  using EosBase<MinimumEnergy<T>>::InternalEnergyFromDensityTemperature;
+  using EosBase<MinimumEnergy<T>>::PressureFromDensityTemperature;
+  using EosBase<MinimumEnergy<T>>::PressureFromDensityInternalEnergy;
+  using EosBase<MinimumEnergy<T>>::MinInternalEnergyFromDensity;
+  using EosBase<MinimumEnergy<T>>::EntropyFromDensityTemperature;
+  using EosBase<MinimumEnergy<T>>::EntropyFromDensityInternalEnergy;
+  using EosBase<MinimumEnergy<T>>::SpecificHeatFromDensityTemperature;
+  using EosBase<MinimumEnergy<T>>::SpecificHeatFromDensityInternalEnergy;
+  using EosBase<MinimumEnergy<T>>::BulkModulusFromDensityTemperature;
+  using EosBase<MinimumEnergy<T>>::BulkModulusFromDensityInternalEnergy;
+  using EosBase<MinimumEnergy<T>>::GruneisenParamFromDensityTemperature;
+  using EosBase<MinimumEnergy<T>>::GruneisenParamFromDensityInternalEnergy;
+  using EosBase<MinimumEnergy<T>>::FillEos;
+
+  using BaseType = T;
+
+  // give me std::format or fmt::format...
+  static std::string EosType() {
+    return std::string("MinimumEnergy<") + T::EosType() + std::string(">");
+  }
+
+  static std::string EosPyType() { return std::string("MinimumEnergy") + T::EosPyType(); }
+
+  MinimumEnergy() = default;
+
+  auto GetOnDevice() { return MinimumEnergy<T>(t_.GetOnDevice()); }
+  inline void Finalize() { t_.Finalize(); }
+
+  template <typename Indexer_t = Real *>
+  PORTABLE_FUNCTION Real TemperatureFromDensityInternalEnergy(
+      const Real rho, const Real sie, Indexer_t &&lambda = nullptr) const {
+    const Real min_sie = t_.MinInternalEnergyFromDensity(rho);
+    return t_.TemperatureFromDensityInternalEnergy(rho, std::max(sie, min_sie), lambda);
+  }
+  template <typename Indexer_t = Real *>
+  PORTABLE_FUNCTION Real InternalEnergyFromDensityTemperature(
+      const Real rho, const Real temperature, Indexer_t &&lambda = nullptr) const {
+    return t_.InternalEnergyFromDensityTemperature(rho, temperature, lambda);
+  }
```

Comment:
We may wish to compute t of emin or something for rho e based EOS's. Not sure that's worth thinking about here though.

Reply:
Are you talking about maxing the input temperature _before_ outputting the energy?

I think more generally though if the energy output here violates the minimum given by the `MinInternalEnergyFromDensity()` function, then it's a reflection of an unphysical EOS (i.e. a negative heat capacity). As a result, I'm not sure we want to guard against this.

That said, it's certainly possible to have a negative heat capacity in some fringe cases, but I'm not sure if this is the appropriate way to guard against this. But then again, maybe it's better to guarantee that the output is always bounded by the minimum.

Reply:
> Are you talking about maxing the input temperature before outputting the energy?

That was what I was thinking but actually it probably makes more sense to max it after. Either way it may not be worth worrying about here.